core/rest-client/src/http_client.rs

```diff
@@ -0,0 +1,425 @@
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{error::Error, Query, RestPath};
+use codec::Encode;
+use http::{
+	header::{HeaderName, AUTHORIZATION, CONTENT_LENGTH, CONTENT_TYPE, USER_AGENT},
+	HeaderValue,
+};
+use http_req::{
+	request::{Method, Request},
+	response::{Headers, Response},
+	uri::Uri,
+};
+use log::*;
+use std::{
+	collections::HashMap,
+	str::FromStr,
+	string::{String, ToString},
+	time::Duration,
+	vec::Vec,
+};
+use url::Url;
+
+pub type EncodedBody = Vec<u8>;
+
+/// Simple trait to send HTTP request, based on `http_req`.
+///
+/// Automatically upgrades to TLS in case the base URL contains 'https'
+pub trait SendHttpRequest {
+	fn send_request<U, T>(
+		&self,
+		base_url: Url,
+		method: Method,
+		params: U,
+		query: Option<&Query<'_>>,
+		maybe_body: Option<String>,
+	) -> Result<(Response, EncodedBody), Error>
+	where
+		T: RestPath<U>;
+}
+
+/// HTTP client implementation
+///
+/// wrapper for the `http_req` library that adds the necessary headers and body to a request
+pub struct HttpClient {
+	send_null_body: bool,
+	timeout: Option<Duration>,
+	headers: Headers,
+	authorization: Option<String>,
+}
+
+impl Default for HttpClient {
+	fn default() -> Self {
+		HttpClient {
+			send_null_body: true,
+			timeout: None,
+			headers: Headers::new(),
+			authorization: None,
+		}
+	}
+}
+
+impl HttpClient {
+	pub fn new(
+		send_null_body: bool,
+		timeout: Option<Duration>,
+		headers: Option<Headers>,
+		authorization: Option<String>,
+	) -> Self {
+		HttpClient {
+			send_null_body,
+			timeout,
+			headers: headers.unwrap_or_else(|| Headers::new()),
+			authorization,
+		}
+	}
+
+	/// Set credentials for HTTP Basic authentication.
+	pub fn set_auth(&mut self, user: &str, pass: &str) {
+		let mut s: String = user.to_string();
+		s.push(':');
+		s.push_str(pass);
+		self.authorization = Some(format!("Basic {}", base64::encode(&s)));
+	}
+
+	/// Set HTTP header from string name and value.
+	///
+	/// The header is added to all subsequent GET and POST requests
+	/// unless the headers are cleared with `clear_headers()` call.
+	pub fn set_header(&mut self, name: &'static str, value: &str) -> Result<(), Error> {
+		let header_name = HeaderName::from_str(name).map_err(|_| Error::InvalidValue)?;
+		let value = HeaderValue::from_str(value).map_err(|_| Error::InvalidValue)?;
+
+		add_to_headers(&mut self.headers, header_name, value);
+		Ok(())
+	}
+
+	/// Clear all previously set headers
+	pub fn clear_headers(&mut self) {
+		self.headers = Headers::new();
+	}
+}
+
+impl SendHttpRequest for HttpClient {
+	fn send_request<U, T>(
+		&self,
+		base_url: Url,
+		method: Method,
+		params: U,
+		query: Option<&Query<'_>>,
+		maybe_body: Option<String>,
+	) -> Result<(Response, EncodedBody), Error>
+	where
+		T: RestPath<U>,
+	{
```

Comment:
Main method to send HTTP requests - builds the URL with all request parameters, headers and body